src/run.ml

```diff
@@ -0,0 +1,497 @@
+    let state, (_ : unit option) = run_as_prover sub_prover_state state in
+    (state, y)
+
+  type response = Request.response
+
+  let unhandled = Request.unhandled
+
+  type request = Request.request =
+    | With :
+        { request: 'a Request.t
+        ; respond: 'a Request.Response.t -> response }
+        -> request
+
+  module Handler = struct
+    type t = request -> response
+  end
+
+  let with_handler ~f h state =
+    let handler = handler state in
+    let state, y = f (set_handler (Request.Handler.push handler h) state) in
+    (set_handler handler state, y)
+
+  let handle f k = with_handler ~f (Request.Handler.create_single k)
+
+  let clear_handler ~f state =
+    let handler = handler state in
+    let state, y = f (set_handler Request.Handler.fail state) in
+    (set_handler handler state, y)
+
+  let exists_provider
+      ~(run : (unit, unit, Field.t, Field.Var.t) Checked.t -> (unit, unit) t)
+      {Types.Typ.store; alloc; check; _} provider state =
+    match prover_state state with
+    | Some s ->
+        let s', value =
+          Provider.run provider (stack state) (get_value state) s
+            (handler state)
+        in
+        let var = Typ_monads.Store.run (store value) (store_field_elt state) in
+        let state = set_prover_state (Some (to_prover_state ())) state in
+        let state, () = run (check var) state in
+        let state = set_prover_state (Some s') state in
+        (state, {Handle.var; value= Some value})
+    | None ->
+        let var = Typ_monads.Alloc.run alloc (alloc_var state) in
+        let state = set_prover_state None state in
+        let state, () = run (check var) state in
+        let state = set_prover_state None state in
+        (state, {Handle.var; value= None})
+
+  let exists ~run ?request ?compute typ =
+    let provider =
+      let request =
+        Option.value request ~default:(As_prover0.return Request.Fail)
+      in
+      match compute with
+      | None -> Provider.Request request
+      | Some c -> Provider.Both (request, c)
+    in
+    bind (exists_provider ~run typ provider) ~f:(fun h -> return (Handle.var h))
+
+  let request_witness ~run
+      (typ : ('var, 'value, Field.t, Field.Var.t) Types.Typ.t)
+      (r : ('value Request.t, Field.Var.t -> Field.t, 's) As_prover0.t) =
+    bind (exists_provider ~run typ (Request r)) ~f:(fun h ->
+        return (Handle.var h) )
+
+  let request ~run ?such_that typ r =
+    match such_that with
+    | None -> request_witness ~run typ (As_prover0.return r)
+    | Some such_that ->
+        let open Let_syntax in
+        let%bind x = request_witness ~run typ (As_prover0.return r) in
+        let%map () = such_that x in
+        x
+
+  let perform ~run req = request_witness ~run (Typ.unit ()) req
+
+  let next_auxiliary () state = (state, !(next_auxiliary state))
+end
+
+module type S_imperative = sig
+  include S with type ('a, _) t = 'a and type 'a prover_state = unit
```

Comment:
As mentioned in our side-conversation, would be good to have in here a field module that actually has operations defined on it. Otherwise you can't really write any programs. I propose we give the user a module here with the signature
```ocaml
module Field : sig
  type t
  val ( * ) : t -> t -> t
  val ( + ) : t -> t -> t
  val ( / ) : t -> t -> t
  val of_string : string -> t
  val of_int : int -> t
end
```
which will be implemented with the type t actually being Cvar.t and the operations being defined using snarky code. E.g., ( * ) will actually be

```ocaml
let ( * ) x y =
  let z = exists typ ~compute:Backend.Field.Infix.( * ) in
  assert_r1cs x y z;
  z
```
where exists and assert_r1cs are the functions in the imperative module.

We should also expose the actual backend field type so that it can be used in as_prover blocks, but I think it should be relegated to a more secondary role (say in Backend.Field or Field.Unchecked or something).